### ISSUE TITLE
Link `node_modules` during platform creation if it exists

### ISSUE THREAD

**Issue by raphinesse** [apache/cordova#32]
After removing bundled `node_modules` from platforms in #6, I noticed that the platforms unconditionally copy their (now potentially missing) bundled `node_modules` in their `create` scripts. Since they all use `shelljs` for that, it currently does not break the creation if `node_modules` but only causes an error message to be displayed.

Example from `cordova-android`:
https://github.com/apache/cordova-android/blob/8dfddef6f9f5d00fba591a2fbe8837ccebea49a7/bin/lib/create.js#L171

A recent conversation with @erisu made me realize that we actually *still* need to either copy or preferably link `<PROJECT>/node_modules/cordova-<PLATFORM>/node_modules` to `<PROJECT>/platforms/<PLATFORM>/node_modules` _if the former exists_. Otherwise Node's module resolution would not pick up the correct dependency versions for the platform in case of version conflicts.

##### Task List
- [x] cordova-android
- [x] cordova-browser
- [x] cordova-ios
- [x] cordova-osx
- [x] cordova-windows
- [x] cordova-test-platform

#### Note
In the long run, we should probably try to have only native projects in `platforms` and keep the platforms' build scripts in the project's `node_modules` and run them from there. That should make things easier and cleaner IMHO.

**Comment by erisu** [comment]
@raphinesse 

1. Is it important/necessary to keep the platform binaries?
```apple_ios_version
apple_ios_version.bat
apple_osx_version
apple_osx_version.bat
apple_xcode_version
apple_xcode_version.bat
autotest
check_reqs
check_reqs.bat
cordova_plist_to_config_xml
create
create.bat
lib
templates
test
tests
testx
uncrustify.cfg
uncrustify.sh
update
update.bat
```

2. Can we remove them? Currently in question is `create`, but the rest can probably go as well.

I believe project creations is typically performed with Cordova CLI. I am not sure how many people actually create a project from the platform package.

# Purpose of the question
## The originating problems background
Once bundled dependencies were removed, when a platform is added, all of the dependencies are installed into the projects root directory, flattened. Therefore the copy command fails because the `node_modules/cordova-ios/node_modules` directory no longer exists.

## Removing the bug
Yes, the best solution is to remove the copy command but there are still test failures to resolve.

## The test failures
The failures are coming from `create.spec.js`, an e2e test file. The test uses the `create` binary to create a project. 

If I attempt to copy the `node_modules` directory into the temp project folder before it is called, create process will fail because it will appear as if the project already exists.

## The question
Basically, there were two options that I can think of. If CLI `cordova platform add ios` is the typical pattern, should the ability to create a project from the platform be valid? If not it can be removed.

If we want to maintain this option, we can update the binary to pass in a flag within the options argument. The flag could be called `isNotCli: true`, and then update the copy node_modules line to come from the `__dirname` and only called when `isNotCli: true`. The reason for a negated option is so we do not need to update CLI code. 

Alternative flags could be: `isPlatformBin: true`, ... 

I am thinking the binaries can be removed and the tests using the binaries as well can be removed.

If you have any confusion let me know and I will try to clear it up.

**Comment by raphinesse** [comment]
@erisu Usually I'm all in favor of reducing the surface of Cordova's public API. However, there are people that actually use platforms on their own (see done of the old open PRs in cordova-lib) and while I myself never have users it before, this actually seems like a good feature to me.

In any case, I think it would be best to not remove any more stuff in next major.

That's all just a first impression. I really need to take a closer look to make an informed decision. I hope I'll be able to do that this weekend.

**Comment by dpogue** [comment]
An interesting problem (that we can defer for the moment) is that our dist packages won't have bundled node_modules anymore, so they won't run out-of-the-box.

Our release process is to generate tarballs and vote on them, and then publish those to npm, so we don't really want the node_modules folder to be in those packages at publish time, but we do want node_modules in the tarball that we make available for download from the website.

**Comment by janpio** [comment]
Is the tarball really a fully a supported distribution method (besides npm)? Does anybody use that? Why?

**Comment by dpogue** [comment]
To work a project that consists of a single platform, using the `./bin/create` script from the tarball instead of cordova-cli.

**Comment by raphinesse** [comment]
From the description of this issue:

> we actually _still_ need to either copy or preferably link `<PROJECT>/node_modules/cordova-<PLATFORM>/node_modules` to `<PROJECT>/platforms/<PLATFORM>/node_modules` _if the former exists_. Otherwise Node's module resolution would not pick up the correct dependency versions for the platform in case of version conflicts.

The recent PRs by @erisu seem to ignore this fact. This is mandatory to not break module resolution. Or did you consider this and I am mistaken here? IMO, we need something like this in the create scripts:

```js
// ROOT is the platform module root
const platformDependencies = path.join(ROOT, 'node_modules');

if (fs.existsSync(platformDependencies)) {
  const localPlatformDependencies = path.join(project_path, 'cordova/node_modules');
  fs.ensureSymlinkSync(platformDependencies, localPlatformDependencies, 'junction');
}
```

That should fix things until we find time to do it the right way (just my opinion):
> In the long run, we should probably try to have only native projects in `platforms` and keep the platforms' build scripts in the project's `node_modules` and run them from there. That should make things easier and cleaner IMHO.

**Comment by janpio** [comment]
Currently installing a platform from a local folder (e.g. `cordova platform add "C:\Projects\cordova5\cordova-android" --verbose`) is broken because of this subject:
https://github.com/apache/cordova-cli/issues/362#issuecomment-443547612

**Comment by erisu** [comment]
Can this ticket be closed or are there still edge cases missing?

**Comment by raphinesse** [comment]
I don't really have a good grasp of the current state of things. But if all platforms behave like Android (if `node_modules` exists then copy it to platforms folder) I believe we are fine for the upcoming major release.

---

Adding a platform straight from a tarball is probably still broken for any platforms without bundled dependencies. However, I believe we should not support that workflow anymore. But that should be discussed in another issue targeting a future major release.

**Comment by erisu** [comment]
@raphinesse If I am correct, you're referring to the standalone non-CLI usage workflow? 

If I am correct, I believe we mentioned somewhere that an extra step to the current workflow is now required.

For users who use the tarball or GH repo for the standalone approach (non-CLI) must run `npm i` before running the `bin/create`. Then the create script would copy the `node_modules` folder into the platform folder as it has done in the past.

_Side note: Preferably to use NPM packaged tarball from the Apache Distribution. It is not recommended to use the GH repo or the GH release package tar/zip as it may contain the package-lock.json. The Apache Distribution package is more accurate and will never have the package-lock.json and should not contain non-essential files._
* [Apache Cordova Platform Release Distribution](https://dist.apache.org/repos/dist/release/cordova/platforms/)
